### ISSUE TITLE
Which browsers and environments would we support

### ISSUE THREAD

**Issue by kuraga** [snabbdom/snabbdom#552]
Seems like there will be a breaking change/major version of Snabbdom. Which browsers would we support? Let's discuss.
/cc @paldepind , @mightyiam , @nojvek

**Comment by mightyiam** [comment]
How about [`defaults`](https://browserl.ist/?q=defaults)?

**Comment by nojvek** [comment]
Care to explain what would be the breaking change ?

I’m of the opinion to support browsers / environments that make at-least 99% of market share + at-least 2 years older versions of mainstream browsers (I think browserlist node plugin allows us to encode this kind of rules easily)

**Comment by kuraga** [comment]
@nojvek We're adding some code to fix IE11 behavior. But do we support IE11? I don't see this information in Read me. So we may
1. to decide which browsers do we support,
2. document it,
3. (if needed/have possibility/not yet done) add testing of these platforms to CI,
4. close all PRs [brings fixes only for] outside of these environments and (start to) remove code [brings fixes only for] outside of these environments.

**Comment by paldepind** [comment]
One observation, there are two "parts" of browser support. One of them is how each browser supports JavaScript, in that respect I think Snabbdom should work in all browsers. The other is how each browser supports the DOM. For the latter, I think it might be ok if some DOM related features in some modules does not work in very old browsers, but the core definitely should.

Another proposal could be that we support all browsers that [support JS's `Set`](https://caniuse.com/#search=new%20Set). That would pretty much only mean dropping support for IE 9 and IE 10. I'm suggesting that only because I have an idea for a potential optimization that makes use of `Set` and which I'll get around to implement and benchmark at some point.

@nojvek I believe #549 is what is being referred to. See also my comment [here](https://github.com/snabbdom/snabbdom/pull/515#issuecomment-575276260). Feedback on both of these is welcome 😄

I'm tagging @staltz and @jvanbruegge as well as this may be relevant to CycleJS.

**Comment by jvanbruegge** [comment]
IE 10 and 9 can still work with a `Set` polyfill. That's how we do it. Officially IE 10 and 9 are not supported but we run at least IE10 with browserstack and some polyfills and it still works.

**Comment by kuraga** [comment]
(about polyfills) Well another question, then: does Snabbdom or end user (application's programmer) add these polyfills? Do we document this or not? (Well obvious questions maybe, but we have to record them, IMHO.)

P.S.
> Care to explain what would be the breaking change ?

I just meant third paragraph of [this comment](https://github.com/snabbdom/snabbdom/pull/515#issuecomment-575276260): @paldepind suggested a major version release.

**Comment by mreinstein** [comment]
please do not break ie support. As much as I hate that piece-of-shit browser, I still have to support it. This library works out of the box with ie10 and up.

**Comment by jvanbruegge** [comment]
@kuraga the end user would end them, or you would force every user to bundle the polyfills even when not needed. And of course this would be documented

**Comment by mreinstein** [comment]
The thing with snabbdom is, yes there are ongoing maintenance tasks and upkeep, but the library is _done_.  It's tiny, it's stable, it has an elegant minimal api, it's performant, it works practically everywhere. 

I'm all for updating to newer tech when there's a clear benefit (for example async/await is a real game changer over tons of nested callbacks.) But I just don't see what kind of benefits we get by ripping out a few lines of array handling with `Set` usage, in exchange for having to load polyfills.

**Comment by kuraga** [comment]
Well then I close this with an appeal to document which browser do we support (and how to cook - polyfills, etc.) Thanks.

**Comment by kuraga** [comment]
@mreinstein but I can't agree with your last comment.
> but the library is done.

No: ![image](https://user-images.githubusercontent.com/1063219/72679899-f06e8f80-3ac4-11ea-92aa-f1f3fc32b9b8.png)
And some of them has been read and labeled after October'19.

> it has an elegant minimal api,

It has the most elegant, most minimal, (al)most modular api, it's the one library (as I understand) with O(1) patch algorithm (thanks, @paldepind !). But some issues show we could do it more elegant...

> it works practically everywhere.

But it doesn't work in modern environment: ECMAScript modules. Cause our style (e.g. `.js` instead of `.mjs` for ECMAScript modules build) is incorrect. So it doesn't work for me.

0.7.3 is stable. It _will be untouched_. This fact doesn't say that 1.0.0 couldn't be another.

P.S. I'm not a Snabbdom developer. Maybe I shouldn't say so. I just wanted to say: 1) thanks!!! (most elegant, ...); 2) it doesn't work with ECMAScript modules; 3) there some issues better to solve before p.2, IMHO.

Thanks!

**Comment by mightyiam** [comment]
I would actually like this to remain open, with a slight change of title, if no-one minds.

**Comment by nojvek** [comment]
Here’s my 2 cents. 

If we can improve the perf / decrease memory usage for the 99% by using better data structures (Map/Set) then we should do so. They are quite optimized under the hood. We should profile our use case to ensure there is a significant benefit.

**Comment by mreinstein** [comment]
So just to clarify, I'm not saying we should avoid changing snabbdom.  Obviously there are always more things to do. Especially around maintaining and doing incremental improvements. 

@kuraga 👍 
> But it doesn't work in modern environment: ECMAScript modules

I agree with you here, and I think this is a good example of a maintenance task for snabbdom. The most current snabbdom release use common js exports, and these work in 0 browsers. So there's already an expectation that one must use a bundler with snabbdom. We could modify snabbom to go from commonjs -> es modules and this wouldn't impact browser support, it would just mean using more modern bundling techniques, which I think is worthwhile too.

**Comment by mreinstein** [comment]
@nojvek 

> We should profile our use case to ensure there is a significant benefit

I agree with you that profiling our use case could yield more insight into what specific benefits `Set` etc. might provide. I've been operating on the assumption that a native js array is not significantly different from `Set` usage in terms of memory, speed, and space.

I guess maybe the harder decision is what values would it take to make requiring polyfills worthwhile. That's more of an opinion (and up to mightyiam and paldepined ultimately.)

e.g., for my personally, if it turned out that Set usage decreases memory by 10%, that would not be worthwhile if I have to transform code to work in the browsers I care about.  I guess we'd also have to measure how much memory snabbdom typically consumes for X number of virtual nodes. I've been assuming that it's already quite memory efficient, but I recognize that's another assumption.

**Comment by paldepind** [comment]
> It has the most elegant, most minimal, (al)most modular api, it's the one library (as I understand) with O(1) patch algorithm (thanks, @paldepind !). But some issues show we could do it more elegant...

Patching in O(1) is not possible. That means constant time. Any patch algorithm would have to be at least `O(n)` where `n` is the size or number of changes that a patch must cause to the DOM.

**Comment by paldepind** [comment]
Per this comment https://github.com/snabbdom/snabbdom/pull/555#issuecomment-580395172 I think that we should keep supporting IE 10.

**Comment by jvanbruegge** [comment]
Yes, requiring polyfills for IE10 to work is fine, but it should be supported

**Comment by mreinstein** [comment]
Maybe I just lack context but I'm assuming there is not much maintenance burden to continue support 
these older browsers.

It works _today_, as is, in older browsers, without making any special build accomodations.

I'm not trying to pick a fight, just want to understand what tangible benefits we get in modifying snabbdom so that one must load polyfills to make it work in older browsers.

**Comment by jvanbruegge** [comment]
No sure of course, but using stuff like `Set` or `Map` might bring a performance boost

**Comment by mreinstein** [comment]
I would definitely be keen to see benchmarks around this, it's a good question. I definitely don't want to penalize 95% of the browsers for the old versions of ie.

**Comment by kuraga** [comment]
> > It has the most elegant, most minimal, (al)most modular api, it's the one library (as I understand) with O(1) patch algorithm (thanks, @paldepind !). But some issues show we could do it more elegant...
>
> Patching in O(1) is not possible. That means constant time. Any patch algorithm would have to be at least O(n) where n is the size or number of changes that a patch must cause to the DOM.

Sure, O(n). Thanks!
